File: python/src/main/python/Q023.py
import heapq

from src.main.python.Solution import Solution
from src.main.python.datastructures.ListNode import ListNode

# Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.


class Q023(Solution):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        heap = []
        for list in lists:
            if list:
                heap.append((list.val, list))
        heapq.heapify(heap)
        head = ListNode(0)
        curr = head
        while heap:
            pop = heapq.heappop(heap)
            curr.next = ListNode(pop[0])
            curr = curr.next
            if pop[1].next:
                heapq.heappush(heap, (pop[1].next.val, pop[1].next))
        return head.next

File: java/src/main/java/solutions/Q023.java
package solutions;

import solutions.datastructures.ListNode;

import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
 */

public class Q023 extends Solution {

    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) return null;

        PriorityQueue<ListNode> queue= new PriorityQueue<>(lists.length, new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1,ListNode o2) { return o1.val - o2.val; }
        });

        ListNode dummy = new ListNode(0), tail = dummy;

        for (ListNode node : lists) {
            if (node != null) {
                queue.offer(node);
            }
        }

        while (!queue.isEmpty()) {
            tail.next = queue.poll();
            tail = tail.next;

            if (tail.next != null) {
                queue.offer(tail.next);
            }
        }

        return dummy.next;
    }

}
